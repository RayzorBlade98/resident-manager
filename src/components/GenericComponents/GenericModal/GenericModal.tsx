/* eslint-disable react/jsx-no-useless-fragment */
import {
  MDBBtn,
  MDBModal,
  MDBModalDialog,
  MDBModalContent,
  MDBModalHeader,
  MDBModalTitle,
  MDBModalBody,
  MDBModalFooter,
} from 'mdb-react-ui-kit';
import React from 'react';

export interface GenericModalProps {
  show: boolean;
  onClose: () => void;
  title: string;
  size?: 'sm' | 'lg' | 'xl';
}

export function GenericModal(
  props: React.PropsWithChildren<GenericModalProps>,
): JSX.Element {
  const children = React.Children.toArray(props.children);
  if (children.length !== 2) {
    throw new Error(
      'Modal needs exactly two children: The body and footer content',
    );
  }
  return (
    <>
      {props.show && (
        <MDBModal show tabIndex="-1" staticBackdrop>
          <MDBModalDialog size={props.size}>
            <MDBModalContent>
              <MDBModalHeader>
                <MDBModalTitle>{props.title}</MDBModalTitle>
                <MDBBtn
                  className="btn-close"
                  color="none"
                  onClick={props.onClose}
                />
              </MDBModalHeader>
              <MDBModalBody>{children[0]}</MDBModalBody>
              <MDBModalFooter>{children[1]}</MDBModalFooter>
            </MDBModalContent>
          </MDBModalDialog>
        </MDBModal>
      )}
    </>
  );
}

export default GenericModal;
