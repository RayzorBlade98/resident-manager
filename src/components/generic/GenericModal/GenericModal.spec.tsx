/* eslint-disable max-len */

import {
  RenderResult,
  cleanup,
  fireEvent,
  render,
} from '@testing-library/react';
import React from 'react';
import GenericModal from '_/components/generic/GenericModal/GenericModal';
import RecoilTestWrapper from '_tests/__test_utils__/RecoillTestWrapper';

describe('GenericModal', () => {
  let renderResult: RenderResult;
  let onModalCloseCallback: jest.Mock;

  function setupRendering(show = true, includeContent = true): void {
    onModalCloseCallback = jest.fn();
    renderResult = render(
      <RecoilTestWrapper>
        <GenericModal
          show={show}
          onClose={onModalCloseCallback}
          title="Test Title"
        >
          {includeContent && <div className="test-body" />}
          {includeContent && <div className="test-footer" />}
        </GenericModal>
      </RecoilTestWrapper>,
    );
  }

  beforeEach(() => {
    setupRendering();
  });

  test('Should call onClose when clicking the exit button', () => {
    // Act
    const exitButton = renderResult.getByRole('button');
    fireEvent.click(exitButton);

    // Assert
    expect(onModalCloseCallback).toHaveBeenCalledTimes(1);
  });

  test('Should throw error if no children are provided', () => {
    // Act
    let error = false;
    try {
      setupRendering(true, false);
    } catch (e) {
      error = true;
    }

    // Assert
    expect(error).toBeTruthy();
  });

  test('should not render closed modal', () => {
    // Arrange
    cleanup();
    setupRendering(false);

    // Assert
    const modal = renderResult.queryByRole('dialog');
    expect(modal).toBeNull();
  });

  test('should match snapshot', () => {
    // Assert
    expect(renderResult.baseElement).toMatchSnapshot();
  });
});
