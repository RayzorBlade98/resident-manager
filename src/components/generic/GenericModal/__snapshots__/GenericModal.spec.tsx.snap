// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericModal should match snapshot 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDialog-root MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
      role="presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-labelledby=":r4:"
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm css-1t1j96h-MuiPaper-root-MuiDialog-paper"
        role="dialog"
      >
        <h2
          class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-bdhsul-MuiTypography-root-MuiDialogTitle-root"
          id=":r4:"
        >
          Test Title
          <button
            aria-label="close"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-lc2yqr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </h2>
        <div
          class="MuiDialogContent-root css-uslfu7-MuiDialogContent-root"
        >
          <div
            class="test-body"
          />
        </div>
        <div
          class="MuiDialogActions-root MuiDialogActions-spacing css-hlj6pa-MuiDialogActions-root"
        >
          <div
            class="test-footer"
          />
        </div>
      </div>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
